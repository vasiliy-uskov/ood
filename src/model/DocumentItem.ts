import {Image, ReadonlyImage} from "./element/Image";
import {Paragraph} from "./element/Paragraph";

export enum DocumentItemType {
	IMAGE = 'image',
	PARAGRAPH = 'paragraph',
}

type ContentItemType = Image|Paragraph

export class DocumentItem {
	private constructor(itemType: DocumentItemType, item: ContentItemType) {
		this._item = item;
		this._itemType = itemType;
	}

	toParagraph(): Paragraph | null {
		if (this._itemType === DocumentItemType.PARAGRAPH)
		{
			return this._item as Paragraph;
		}
		return null
	}

	toImage(): Image | null {
		if (this._itemType === DocumentItemType.IMAGE)
		{
			return this._item as Image;
		}
		return null
	}

	static fromImage(image: Image) {
		return new DocumentItem(DocumentItemType.IMAGE, image);
	}

	static fromParagraph(paragraph: Paragraph) {
		return new DocumentItem(DocumentItemType.PARAGRAPH, paragraph);
	}


	private readonly _itemType: DocumentItemType;
	private readonly _item: ContentItemType;
}

export interface ReadonlyDocumentItem {
	toParagraph(): Paragraph | null;
	toImage(): ReadonlyImage | null;
}