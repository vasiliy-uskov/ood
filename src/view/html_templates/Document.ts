import {escapeHtmlText} from "../../utils/utils";
import {Image, ImageProps} from "./Image";
import {Paragraph, ParagraphProps} from "./Paragraph";

export enum ContentType {
	PARAGRAPH = 'paragraph',
	IMAGE = 'image',
}

export type ParagraphDocumentItem = {
	type: ContentType.PARAGRAPH,
	props: ParagraphProps,
}

export type ImageDocumentItem = {
	type: ContentType.IMAGE,
	props: ImageProps,
}

export type DocumentContent = ReadonlyArray<(ParagraphDocumentItem|ImageDocumentItem)>;

export type DocumentProps = {
	title: string,
	content: DocumentContent,
}

export function Document(props: DocumentProps) {
	return `<!DOCTYPE html>
<html lang="en">
<head>
	<meta charset="UTF-8">
	<title>${escapeHtmlText(props.title)}</title>
</head>
<body>
	${props.content.map(item => {
		switch (item.type) {
			case ContentType.IMAGE:
				return Image(item.props);
			case ContentType.PARAGRAPH:
				return Paragraph(item.props);
			default:
				throw new Error("Undefined content type")
		}
	}).join("")}
</body>
</html>`;
}