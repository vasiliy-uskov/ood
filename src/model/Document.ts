import {DocumentItem, ReadonlyDocumentItem} from "./DocumentItem";

export class Document {
	public readonly content: ReadonlyArray<DocumentItem>;
	public readonly title: string;
}

export class ReadonlyDocument {
	public readonly content: ReadonlyArray<ReadonlyDocumentItem>;
	public readonly title: string;
}