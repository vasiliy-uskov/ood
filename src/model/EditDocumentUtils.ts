import {DocumentItem} from "./DocumentItem";
import {Document} from "./Document";

export function insertItem(document: Document, item: DocumentItem, position: number): Document|null {
	if (position < 0 || document.content.length < position)
	{
		return null
	}
	const content = document.content.slice();
	content.splice(position, 0, item);
	return {
		...document,
		content,
	}
}

export function deleteItem(document: Document, position: number): Document|null {
	if (!document.content.length || position < 0 || document.content.length <= position)
	{
		return null;
	}
	const content = document.content.slice();
	content.splice(position, 1);
	return {
		...document,
		content
	}
}

export function modifyItem(document: Document, position: number, replaceFn: (item: DocumentItem) => DocumentItem): Document|null {
	if (!document.content.length || position < 0 || document.content.length <= position)
	{
		return null;
	}
	const content = document.content.slice();
	const item = content[position];
	content.splice(position, 1, replaceFn(item));
	return {
		...document,
		content
	}
}