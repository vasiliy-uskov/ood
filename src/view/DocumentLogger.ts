import {ReadonlyDocument} from "../model/Document";
import {ReadonlyImage} from "../model/element/Image";
import {Paragraph} from "../model/element/Paragraph";

export class DocumentLogger {
	public static log(document: ReadonlyDocument): void {
		this._logDocument(document)
	}

	private static _logDocument(document: ReadonlyDocument) {
		console.log('Title: ', document.title);
		document.content.forEach((item, index) => {
			const image = item.toImage();
			const paragraph = item.toParagraph();
			if (image)
			{
				this._logImage(image, index);
			}
			if (paragraph)
			{
				this._logParagraph(paragraph, index);
			}
		})
	}

	private static _logParagraph(paragraph: Paragraph, index: number) {
		console.log(`${index}. Paragraph: ${paragraph.text}`)
	}

	private static _logImage(image: ReadonlyImage, index: number) {
		console.log(`${index}. Image: ${image.size.width} ${image.size.height} ${image.img.srcFileName()}`)
	}
}