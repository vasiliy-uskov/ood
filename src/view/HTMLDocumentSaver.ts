import {ReadonlyDocument} from "../model/Document";
import {writeFileSync} from "fs";
import {join, relative} from "path";
import {DocumentProps, Document, ContentType, ParagraphDocumentItem, ImageDocumentItem} from "./html_templates/Document";
import {ReadonlyImage} from "../model/element/Image";
import {Paragraph} from "../model/element/Paragraph";

const DOCUMENT_FILE_NAME = 'index.html';
const IMAGES_DIR_NAME = 'images';

export class HTMLDocumentSaver {
	public static save(location: string, document: ReadonlyDocument): void {
		const destIndexPath = join(location, DOCUMENT_FILE_NAME);
		writeFileSync(destIndexPath, Document(this._processResource(document, location)))
	}

	private static _processResource(document: ReadonlyDocument, location: string): DocumentProps {
		return {
			title: document.title,
			content: document.content.map(item => {
				const image = item.toImage();
				const paragraph = item.toParagraph();
				if (image)
				{
					return this._processImage(image, location)
				}
				if (paragraph)
				{
					return this._processParagraph(paragraph);
				}
				throw new Error();
			})
		}
	}

	private static _processImage(image: ReadonlyImage, location: string): ImageDocumentItem {
		const imagesPath = join(location, IMAGES_DIR_NAME);
		const src = relative(location, image.img.save(imagesPath));
		return {
			type: ContentType.IMAGE,
			props: {
				size: image.size,
				src,
			}
		};
	}

	private static _processParagraph(paragraph: Paragraph): ParagraphDocumentItem {
		return {
			type: ContentType.PARAGRAPH,
			props: paragraph,
		};
	}
}