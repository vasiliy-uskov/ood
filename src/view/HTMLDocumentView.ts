import {ReadonlyDocument} from "../model/Document";
import {ReadonlyImage} from "../model/element/Image";
import {Paragraph} from "../model/element/Paragraph";
import {writeFileSync} from "fs";
import {join, relative} from "path";

const DOCUMENT_FILE_NAME = 'index.html';
const IMAGES_DIR_NAME = 'images';

export class HTMLDocumentView {
	constructor(destPath: string) {
		this._destPath = destPath;
		this._imagesPath = join(this._destPath, IMAGES_DIR_NAME);
	}

	draw(document: ReadonlyDocument): void {
		writeFileSync(
			join(this._destPath, DOCUMENT_FILE_NAME),
			this._expandDocument(document)
		)
	}

	private _expandDocument(document: ReadonlyDocument): string {
		return `<!DOCTYPE html>
<html lang="en">
<head>
	<meta charset="UTF-8">
	<title>${this._escapeText(document.title)}</title>
</head>
<body>
	${document.content.map(item => {
			const image = item.toImage();
			const paragraph = item.toParagraph();
			if (image)
			{
				return this._expandImage(image);
			}
			if (paragraph)
			{
				return this._expandParagraph(paragraph);
			}
			return '';
	}).join('')}
</body>
</html>`;
	}

	private _expandParagraph(paragraph: Paragraph) {
		return `<p>${this._escapeText(paragraph.text)}</p>`;
	}

	private _expandImage(image: ReadonlyImage): string {
		const src = relative(this._destPath, image.img.save(this._imagesPath));
		return `<img
					src="${src}"
					width="${image.size.width}"
					height="${image.size.height}"
					alt=""
				/>`
	}

	private _escapeText(text: string) {
		return text
			.replace(/&/g,'&amp;')
			.replace(/</g,'&lt;')
			.replace(/>/g,'&gt;')
	}

	private readonly _destPath: string;
	private readonly _imagesPath: string;
}