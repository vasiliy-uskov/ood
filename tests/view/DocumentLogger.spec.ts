import {DocumentLogger} from "../../src/view/DocumentLogger";
import {createParagraph, createStaticNameImage} from "../mocks/DocumentData";
import {DocumentItem} from "../../src/model/DocumentItem";
import {ConsoleLogMock} from "../mocks/ConsoleLogMock";
import {ReadonlyDocument} from "../../src/model/Document";

function logDocument(document: ReadonlyDocument) {
	ConsoleLogMock.startMock();
	DocumentLogger.log(document);
	expect(ConsoleLogMock.result()).toMatchSnapshot();
	ConsoleLogMock.stopMock();
}

it('log document with content', () => {
	logDocument({
		title: 'Some title',
		content: [
			DocumentItem.fromParagraph(createParagraph('Some text')),
			DocumentItem.fromImage(createStaticNameImage('some_image', {
				width: 100,
				height: 100,
			}))
		]
	});
});

it('log document with title and without content', () => {
	logDocument({
		title: 'Some title',
		content: []
	});
});

it('log empty document', () => {
	logDocument({
		title: '',
		content: []
	});
});