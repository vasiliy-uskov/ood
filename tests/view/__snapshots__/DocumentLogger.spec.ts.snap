// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`log document with content 1`] = `
"Title: Some title
0. Paragraph: Some text
1. Image: 100 100 some_image.mock
"
`;

exports[`log document with title and without content 1`] = `
"Title: Some title
"
`;

exports[`log empty document 1`] = `
"Title: 
"
`;
